google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/functions/WindowsPasswordGenerator.java

```diff
@@ -0,0 +1,241 @@
+ * <ol>
+ *   <li>Generate a temporary key for encrypting and decrypting the password
+ *   <li>Send the RSA public key to the instance, by settings its metadata
+ *   <li>Retrieve the result from the {@link SerialPortOutput}
+ *   <li>Decode and decrypt the result.
+ * </ol>
+ */
+public class WindowsPasswordGenerator implements Function<Map<String, ?>, String> {
+
+   /**
+    * Indicates when the key should expire. Keys are one-time use, so the metadata doesn't need to stay around for long.
+    * 10 minutes chosen to allow for differences between time on the client
+    * and time on the server.
+    */
+   private static final long EXPIRE_DURATION = 10 * 60 * 1000;
+   private String userName;
+   private String email;
+
+    @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final GoogleComputeEngineApi api;
+   private final Crypto crypto;
+   private final Predicate<AtomicReference<Operation>> operationDone;
+   
+   @Inject
+   protected WindowsPasswordGenerator(GoogleComputeEngineApi api, Crypto crypto, Predicate<AtomicReference<Operation>> operationDone) {
+      this.api = api;
+      this.crypto = crypto;
+      this.operationDone = operationDone;
+   }
+
+   @Override
+   public String apply(Map<String, ?> params) {
+      String zone = (String)params.get("zone");
+       final AtomicReference<Instance> instance = (AtomicReference<Instance>)params.get("instance");
+       userName = (String)params.get("userName");
+       email = (String)params.get("email");
+
+      // TODO Check whether VM is up
+      try {
+         // Generate the public/private key pair for encryption and decryption.
+         // TODO do we need to explicitly set 2048 bits? Presumably "RSA" is implicit
+         KeyPair keys = crypto.rsaKeyPairGenerator().genKeyPair();
+
+         // Update instance's metadata with new "windows-keys" entry, and wait for operation to 
+         // complete.
+         logger.debug("Generating windows key for instance %s, by updating metadata", instance.get().name());
+         final InstanceApi instanceApi = api.instancesInZone(zone);
+         Metadata metadata = instance.get().metadata();
+         // If disableHtmlEscaping is not there, == will be escaped from modulus value
+          metadata.put("windows-keys", new GsonBuilder().disableHtmlEscaping().create().toJson(extractKeyMetadata(keys)));
+
+         AtomicReference<Operation> operation = Atomics.newReference(instanceApi.setMetadata(instance.get().name(), metadata));
+         operationDone.apply(operation);
+
+         if (operation.get().httpErrorStatusCode() != null) {
+            logger.warn("Generating windows key for %s failed. Http Error Code: %d HttpError: %s",
+                  operation.get().targetId(), operation.get().httpErrorStatusCode(),
+                  operation.get().httpErrorMessage());
+         }
+
+         TimeLimiter limiter = new SimpleTimeLimiter();
```

Comment:
Andrea's suggestion is still valid (and preferred). The ``Predicates2.retry` method will return as soon as the predicate succeeds.